Replace copier.Copy with optimized manual deep copy in URL.Clone (#2678)

Nexusrex18 opened the pull request:
Fixes by replacing `copier.Copy` with a manual deep copy in `URL.Clone()`, optimized with pre-allocated maps for `params` and `attributes`. This reduces memory and CPU overhead, especially for large data volumes. 
Key changes:
- Replaced `copier.Copy` with manual implementation in `URL.Clone()`.
- Removed `github.com/jinzhu/copier` import from `url.go` and dependency from `go.mod` via `go mod tidy`.
- Added `TestSubURLCopy` to verify deep-copy behavior.
- Updated `BenchmarkClone` to test the new implementation.

Benchmarks:
- 100 params/attributes:
  - Old (copier): 51 µs/op, 23.74 KB/op, 222 allocs/op
  - New (manual): 26 µs/op, 14.12 KB/op, 112 allocs/op
- 1,000 params/attributes:
  - Old (copier): 428 µs/op, 228 KB/op, 1,142 allocs/op
  - New (manual): 270 µs/op, 197 KB/op, 1,010 allocs/op

The new implementation is faster (~1.6x at 1,000 entries, ~2x at 100), uses less memory (~14% at 1,000, ~40% at 100), and reduces allocations (~12% at 1,000, ~50% at 100).

Note: Unrelated test failures in `rpc_service_test.go` (e.g., unexported `testService`) were observed but pre-exist this change and are outside this PR’s scope.

Fixes: #2678 

Tested with:
- `go test -v -run TestSubURLCopy`: Passed
- `go test -bench=BenchmarkClone -benchmem`: Confirmed performance

No-SilverBullet commented on common/url.go:
```diff
@@ -829,20 +829,35 @@ func (c *URL) MergeURL(anotherUrl *URL) *URL {
+		copy(newValues, values)
+		newURL.params[key] = newValues
+	}
+	c.paramsLock.RUnlock()
```
Use defer, `defer c.paramsLock.Runlock()`，and follow where the lock used(line 845)

No-SilverBullet commented on common/url.go:
```diff
@@ -829,20 +829,35 @@ func (c *URL) MergeURL(anotherUrl *URL) *URL {
+	for key, value := range c.attributes {
+		newURL.attributes[key] = value
+	}
+	c.attributesLock.RUnlock()
```
same with paramslock

No-SilverBullet commented on common/url.go:
```diff
@@ -829,20 +829,35 @@ func (c *URL) MergeURL(anotherUrl *URL) *URL {
+		Methods:      make([]string, len(c.Methods)),
+	}
+	copy(newURL.Methods, c.Methods)
+	newURL.params = make(url.Values, len(c.params))
```
put this line of code in the newURL initialization above, more concise

No-SilverBullet commented on common/url.go:
```diff
@@ -829,20 +829,35 @@ func (c *URL) MergeURL(anotherUrl *URL) *URL {
+		Path:         c.Path,
+		Username:     c.Username,
+		Password:     c.Password,
+		Methods:      make([]string, len(c.Methods)),
```
change to -> `Methods:     make([]string, len(c.Methods)) `, and remove line843 copy, more concise

FoghostCn reviewed (approved):
LGTM

Nexusrex18 commented:
@No-SilverBullet 
I updated URL.Clone() in common/url.go to use defer for lock releases and added nil checks for params and attributes as per the suggestions. Additionally, I enhanced TestSubURLCopy to test Protocol and params for thorough deep-copy validation and fixed TestCompareURLEqualFunc with SetCompareURLEqualFunc(defaultCompareURLEqual) resets to address state pollution.

Nexusrex18 commented:
@No-SilverBullet 
Rebased `issue-2678` onto `upstream/develop` (88823f4a) for CI sync. Benchmarks: 303 µs/op (avg of 5 runs), ~1.41x faster than `copier.Copy`. All `common/` tests pass locally (Windows & Ubuntu WSL, including `-race`), but Ubuntu CI fails: `make: *** [Makefile:59: test] Error 1`. This was fixed in #2795  by `develop`, but persists here.

No-SilverBullet commented:
Have you pull the newest upsteam/develop branch?  CI error still exists.

Nexusrex18 commented:
@No-SilverBullet even after fetching the latest changes , getting the same issue so should i try any other approach or recreate a new pull request with same changes ?
